singularity-eos/eos/eos_variant.hpp

```diff
@@ -229,6 +230,249 @@ class Variant {
+                                                  const int num,
+                                                  RealIndexer &&bmods,
+                                                  LambdaIndexer &&lambdas) const {
+    return mpark::visit(
+        [&rhos, &sies, &num, &bmods, &lambdas](const auto &eos) {
+          return eos.BulkModulusFromDensityInternalEnergyVector(
+              rhos, sies, bmods, lambdas);
+        },
+        eos_);
+  }
+
+  template<typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
+  inline
+  void GruneisenParamFromDensityTemperatureVector(ConstRealIndexer &&rhos,
+                                                  ConstRealIndexer &&temperatures,
+                                                  const int num,
+                                                  RealIndexer &&gm1s,
+                                                  LambdaIndexer &&lambdas) const {
+    return mpark::visit(
+        [&rhos, &temperatures, &num, &gm1s, &lambdas](const auto &eos) {
+          return eos.GruneisenParamFromDensityTemperatureVector(
+              rhos, temperatures, gm1s, lambdas);
+        },
+        eos_);
+  }
+
+  template<typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
+  inline
+  void GruneisenParamFromDensityInternalEnergyVector(ConstRealIndexer &&rhos,
+                                                     ConstRealIndexer &&sies,
+                                                     const int num,
+                                                     RealIndexer &&gm1s,
+                                                     LambdaIndexer &&lambdas) const {
+    return mpark::visit(
+        [&rhos, &sies, &num, &gm1s, &lambdas](const auto &eos) {
+          return eos.GruneisenParamFromDensityInternalEnergyVector(
+              rhos, sies, gm1s, lambdas);
+        },
+        eos_);
+  }
+
+  template<typename RealIndexer, typename LambdaIndexer>
+  inline
+  void FillEosVector(RealIndexer &&rhos, RealIndexer &&temps, RealIndexer &&energies,
+                     RealIndexer &&presses, RealIndexer &&cvs, RealIndexer &&bmods,
+                     const int num, const unsigned long output,
+                     LambdaIndexer &&lambdas) const {
+    return mpark::visit(
+        [&rhos, &temps, &energies, &presses, &cvs, &bmods, &num, &output,
+         &lambdas](const auto &eos) {
+          return eos.FillEosVector(rhos, temps, energies, presses, cvs, bmods,
+                                   num, output, lambdas);
+        },
+        eos_);
+  }
+
+  template<typename RealIndexer, typename LambdaIndexer>
+  inline
+  void PTofREVector(RealIndexer &&rhos, RealIndexer &&sies, const int num,
+                    LambdaIndexer &&lambdas, RealIndexer &&presses,
+                    RealIndexer &&temps, RealIndexer &&dpdrs,
+                    RealIndexer &&dpdes, RealIndexer &&dtdrs,
+                    RealIndexer &&dtdes) const {
+    return mpark::visit(
```

Comment:
Hmm... I'm not sure how I feel about re-implementing this whole thing here. Maybe it's necessary, I dunno. This whole `PTofRE` machinery might be better being refactored to reduced code duplication, and to enable individual EOSs to implement their own version if desired. (Similar to how you did the `Vector` loop abstractions.)

Reply:
I re-implemented it simply because I was adapting what was already there. I tried to mirror the existing implementation as closely as I could just with arrays.

I think my main problem here is that different EOS may want to calculate derivatives differently and the 1e-06 factor for the perturbation is just an arbitrary number that may or may not be sufficient for all situations. I'd propose keeping the functionality similar between the scalar and vector versions of these functions, but I'm otherwise open to suggestions for to change this.

Reply:
My suggestion is we should see if we can make a generic implementation that we call from each EOS, then this can be overloaded as needed.

But we can deal with that in a later PR.